content/browser/attribution_reporting/attribution_storage_sql.cc

```diff
@@ -2999,6 +3388,272 @@ bool AttributionStorageSql::StoreAttributionReport(AttributionReport& report) {
   return true;
 }
 
+AggregatableResult AttributionStorageSql::PayAllOrNothing(
+          attribution_reporting::AttributionWindow attribution_window,
+          const attribution_reporting::SuitableOrigin& querying_origin,
+          double required_budget) {
+
+  std::vector<uint64_t> attribution_epochs;
+  for (uint64_t i=attribution_window.epoch_start(); 
+          i<=attribution_window.epoch_end(); ++i) {
+    attribution_epochs.push_back(i);
+  }
+  return PayAllOrNothing(attribution_epochs, querying_origin, required_budget);
+}
+
+AggregatableResult AttributionStorageSql::PayAllOrNothing(
+          std::vector<uint64_t> attribution_epochs,
+          const attribution_reporting::SuitableOrigin& querying_origin,
+          double required_budget) {
+
+  // Starting atomic operation
+  sql::Transaction transaction(&db_);
+  if (!transaction.Begin()) {
+    return AggregatableResult::kInternalError;
+  }
+
+  // Obtain remaining budget per epoch for all attribution epochs  
+  std::ostringstream oss; 
+  oss << \
+    "SELECT F.epoch, (F.initial_budget - F.consumed_budget) AS remaining_budgets "
+      "FROM per_origin_filters F "
+      "WHERE F.origin=? "
+      "AND F.epoch IN (";
+
+  auto& items = attribution_epochs;
+  if (!items.empty()) {
+    std::copy(items.begin(), items.end() - 1, std::ostream_iterator<uint64_t>(oss, ", "));
+    oss << items.back();
+  }
+  oss << ")";
+
```

Comment:
Wow that's how they do SQL queries inside the browser?? That looks so sketchy to manipulates strings like that. But I guess it's safe

Reply:
No, I do this ugly thing! I want to feed dynamically a variable number of epochs in the string and the template doesn't allow you to do that. At least it's not among the SQL functionality imported in Chrome. 
Theoretically it is supported to bind a '?' symbol with an array (https://www.sqlite.org/carray.html) but this is not imported. I tried to play around a bit to support that but it turned out to be a headache and I did this ugly thing instead and move on.